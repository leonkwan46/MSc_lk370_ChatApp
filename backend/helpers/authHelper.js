import { Parent, Student, Teacher, User } from "../db/modals/index.js"
import bcrypt from 'bcrypt'
import jwt from 'jsonwebtoken'
import userDataHelper from "./userDataHelper.js"

const authHelper = {}

authHelper.validateUserByEmail = async (email, isLogin=false) => {
    let user = await User.findOne({ email })

    if (isLogin) {
        if (!user) throw new Error("Invalid Username/Password")
        if (!user.isRegistered) throw new Error("User not registered")
    } else {
        if (user) throw new Error("User already exists")
    }
    return user
}

authHelper.validatePassword = async (password, hashPassword) => {
    const isMatch = await bcrypt.compare(password, hashPassword)
    if (!isMatch) throw new Error("Invalid Username/Password")
    return isMatch
}

authHelper.returnUserDataToClient = async (user = '', isValid = false) => {
    if (!isValid) user = await authHelper.validateUserByEmail(user.email, true)
    if (!user) throw new Error("User not found")

    const defaultUserData = {
        userId: user._id,
        email: user.email,
        role: user.role,
        isRegistered: user.isRegistered,
        isGeneralFormComplete: user.isGeneralFormComplete,
    }

    let additionalUserData = {}
    switch (user.role) {
        case "parent":
            additionalUserData = await userDataHelper.fetchParentUserData(user)
            break
        case "teacher":
            additionalUserData = await userDataHelper.fetchTeacherUserData(user)
            break
        case "student":
            additionalUserData = await userDataHelper.fetchStudentUserData(user)
            break
        default:
            break
    }

    return { ...defaultUserData, ...additionalUserData }
}

authHelper.generateHashPassword = async (password) => {
    const salt = await bcrypt.genSalt(10)
    return await bcrypt.hash(password, salt)
}

authHelper.generateAuthToken = async (user) => {
    const tokenPayload = {
        _id: user._id,
        email: user.email,
        hashPassword: user.hashPassword,
        role: user.role,
        gender: user.gender,
        name: user.name,
        isRegistered: user.isRegistered,
        isInvited: user.isInvited,
        invitationCode: user.invitationCode,
        isInvitationVerified: user.isInvitationVerified,
        isDocUploaded: user.isDocUploaded,
        isDocVerified: user.isDocVerified,
        isGeneralFormComplete: user.isGeneralFormComplete,
        createAt: user.createAt,
    }
    return await jwt.sign(tokenPayload, process.env.JWT_SECRET, { expiresIn: "1d" })
}

authHelper.createAccount = async (email, password, role) => {
    let user = await authHelper.validateUserByEmail(email)
    const hashPassword = await authHelper.generateHashPassword(password)
    // Create User
    if (role === "parent") {
        user = new Parent({
            email,
            hashPassword,
            role,
            isRegistered: true,
        })
    } else if (role === "teacher") {
        user = new Teacher({
            email,
            hashPassword,
            role,
            isRegistered: true,
        })
    } else if (role === "student") {
        user = new Student({
            email,
            hashPassword,
            role,
            isRegistered: true,
        })
    }
    // Store User
    const storingUser = await user.save()
    if (!storingUser) throw new Error("Failed to store user")

    // Find User
    user = await User.findOne({ email })
    // Generate Token
    const token = await authHelper.generateAuthToken(user)

    return { user, token }
}

export default authHelper