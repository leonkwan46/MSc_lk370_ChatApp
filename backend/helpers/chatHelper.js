import { Message, Room } from '../db/modals/index.js'
import userDataHelper from './userDataHelper.js'

const chatHelper = {}

chatHelper.handleSendMessage = async (io, socket, props) => {
    const { roomId, message, userId, userRole } = props

    const newMessage = Message({
        message,
        senderId: userId,
        senderRole: userRole,
        roomId,
        sentAt: Date.now(),
        isRead: false,
    })

    // Save message to database
    await newMessage.save()

    // Save message to room
    await Room.updateOne({ _id: roomId }, { $push: { messages: newMessage } })

    io.emit('chatMessage', { sender: socket.id, data: newMessage })
}

chatHelper.generateRoomMemberData = async (userID, role, data) => {
    const { teacher, student, parent, child } = data
    let roomMemberData
    switch (role) {
        case 'teacher':
            roomMemberData = {
                teacher: await userDataHelper.fetchUserDataByIDAndRole(userID, 'teacher'),
                student: await userDataHelper.fetchUserDataByIDAndRole(student.split(',')[1], 'student'),
                parent: await userDataHelper.fetchUserDataByIDAndRole(parent.split(',')[1], 'parent')
            }
            break
        case 'parent':
            roomMemberData = {
                parent: await userDataHelper.fetchUserDataByIDAndRole(userID, 'parent'),
                teacher: await userDataHelper.fetchUserDataByIDAndRole(teacher.split(',')[1], 'teacher'),
                student: await userDataHelper.fetchUserDataByIDAndRole(child.split(',')[1], 'student')
            }
            break
        case 'student':
            roomMemberData = {
                student: await userDataHelper.fetchUserDataByIDAndRole(userID, 'student'),
                parent: await userDataHelper.fetchUserDataByIDAndRole(parent.split(',')[1], 'parent'),
                teacher: await userDataHelper.fetchUserDataByIDAndRole(teacher.split(',')[1], 'teacher')
            }
            break
        default:
            break
    }
    roomMemberData

    return roomMemberData
}

chatHelper.joinRoom = async (io, socket, roomData) => {
    const { roomId } = roomData
    const chatHistory = await Room.findById(roomId).populate('messages')

    socket.join(roomId)
    io.to(roomId).emit('roomJoined', { chatHistory })
}


export default chatHelper
