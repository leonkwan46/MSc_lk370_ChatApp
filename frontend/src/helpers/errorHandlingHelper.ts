type ErrorWithMessage = {
    message: string
}

type ErrorWithCustomedMessage = {
    response: {
        data: {
            message: string
        }
    }
}

const isErrorWithMessage = (error: unknown): error is ErrorWithMessage => {
    return (
        typeof error === 'object' &&
        error !== null &&
        'message' in error &&
        typeof (error as ErrorWithMessage).message === 'string'
    )
}

const isErrorWithCustomedMessage = (error: unknown): error is ErrorWithCustomedMessage => {
    return (
        typeof error === 'object' &&
        error !== null &&
        'response' in error &&
        typeof (error as ErrorWithCustomedMessage).response === 'object' &&
        'data' in (error as ErrorWithCustomedMessage).response &&
        typeof ((error as ErrorWithCustomedMessage).response.data) === 'object' &&
        'message' in ((error as ErrorWithCustomedMessage).response.data) &&
        typeof ((error as ErrorWithCustomedMessage).response.data.message) === 'string'
    )
}

const toErrorWithMessage = (maybeError: unknown): ErrorWithMessage | ErrorWithCustomedMessage => {
    if (isErrorWithMessage(maybeError) || isErrorWithCustomedMessage(maybeError)) return maybeError

    try {
        return new Error(JSON.stringify(maybeError))
    } catch {
        // fallback in case there's an error stringifying the maybeError
        // like with circular references for example.
        return new Error(String(maybeError))
    }
}

export const getErrorMessage = (error: unknown) => {
    const errMessage = toErrorWithMessage(error)

    if ('response' in errMessage && 'data' in errMessage.response) return errMessage.response.data.message
    if ('message' in errMessage) return errMessage.message
}

