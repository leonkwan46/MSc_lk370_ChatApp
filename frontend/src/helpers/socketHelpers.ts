import { io, Socket } from 'socket.io-client'
import { Message } from '../types/types'

export interface MessageData {
  id: string
  message: string
  roomId: string
  senderId: string
  senderRole: string
}

const socket: Socket = io('http://localhost:3000')

const getSocketId = (): string => {
  if (!socket.id) throw new Error('Socket ID is undefined')
  return socket.id
}

const disconnectSocket = async (): Promise<void> => {
  await socket.disconnect()
}

const sendMessage = async (roomId: string, message: string, userId: string, userRole: string): Promise<void> => {
  await socket.emit('sendMessage', { roomId, message, userId, userRole })
}

const receiveMessage = (): Promise<MessageData>=> {
  return new Promise((resolve) => {
    socket.on('chatMessage', (res) => {
      resolve(res.data)
    })
  })
}

const joinRoomAndGetChatHistory = (roomId: string): Promise<Message[]> => {
  return new Promise((resolve, reject) => {
      socket.emit('joinRoom', { roomId })
      socket.once('roomJoined', (result) => {
        console.log(result)
          resolve(result.chatHistory)
      })
      socket.once('roomJoinError', (error) => {
          reject(error)
      })
  })
}

export { disconnectSocket, getSocketId, receiveMessage, sendMessage, joinRoomAndGetChatHistory }
