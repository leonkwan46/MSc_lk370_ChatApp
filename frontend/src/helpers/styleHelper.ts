const getColor = (color: string) => {
    switch (color) {
        case 'primary':
            return '#D4AF37'
        case 'secondary':
            return 'white'
        case 'error':
            return 'red'
        case 'warning':
            return 'orange'
        case 'success':
            return 'green'
        case 'info':
            return 'black'
        case 'invisible':
            return 'transparent'
        case 'red':
            return '#D44343'
        case 'green':
            return '#39A139'
        default:
            return 'black'
    }
}

const getFontSize = (size: string) => {
    switch (size) {
        case 'xs':
            return 5
        case 'sm':
            return 10
        case 'md':
            return 15
        case 'lg':
            return 20
        case 'xl':
            return 25
        case 'xxl':
            return 35
        case 'title':
            return 45
        case 'subtitle':
            return 25
        default:
            return 15
    }
}

interface getTextInputStyleProps {
    size: string
    hasError?: boolean
    outline?: boolean
}

const getTextInputStyle = ({ size, hasError=false, outline=false}: getTextInputStyleProps) => {
    const sizes = {
        xs: { height: 25, padding: 5, fontSize: 9 },
        sm: { height: 30, padding: 7, fontSize: 10 },
        md: { height: 40, padding: 8, fontSize: 11 },
        lg: { height: 60, padding: 12, fontSize: 12 },
        xl: { height: 70, padding: 15, fontSize: 13 },
        default: { height: 50, padding: 10, fontSize: 12 }
    }

    const defaultSize = {
        backgroundColor: '#fff',
        borderRadius: 10,
        fontFamily: 'Lemon-Regular',
        width: '100%' as '100%',
        ...sizes.default
    }

    const errorStlye = {
        borderColor: '#f00',
        borderWidth: 2,
    }

    const outlineStyle = {
        borderColor: '#000',
        borderWidth: 2,
    }

    const selectedSize = sizes[size as keyof typeof sizes] || sizes.default

    return {
        ...defaultSize,
        ...selectedSize,
        ...(hasError ? errorStlye : {}),
        ...(outline ? outlineStyle : {})
    }
}

interface getButtonStyleProps {
    size: string
    color: string
    fill?: boolean
}

const getButtonStyle = ({ size, color, fill=true }: getButtonStyleProps) => {
    const sizes = {
        xs: { width: '25%', height: 25, paddingVertical: 5, paddingHorizontal: 10 },
        sm: { width: '25%', height: 30, paddingVertical: 7, paddingHorizontal: 12 },
        md: { height: 40, paddingVertical: 8, paddingHorizontal: 15 },
        lg: { height: 50, paddingVertical: 12, paddingHorizontal: 20 },
        xl: { height: 60, paddingVertical: 15, paddingHorizontal: 25 },
        default: { height: 45, paddingVertical: 10, paddingHorizontal: 18 }
    }

    const colors = getColor(color)

    const defaultContainerStyle = {
        backgroundColor: colors,
        borderColor: '',
        borderRadius: 5,
        borderWidth: 0,
        justifyContent: 'center',
        alignItems: 'center',
        width: '50%' as '50%'
    }

    if (!fill) {
        defaultContainerStyle.backgroundColor = getColor('invisible')
        defaultContainerStyle.borderColor = colors
        defaultContainerStyle.borderWidth = 2
        defaultContainerStyle.borderRadius = 10
    }

    const selectedSize = sizes[size as keyof typeof sizes] || sizes.default

    return {
        containerStyle: {
            ...defaultContainerStyle,
            ...selectedSize,
        }
    }
}





export { getColor, getFontSize, getTextInputStyle, getButtonStyle }