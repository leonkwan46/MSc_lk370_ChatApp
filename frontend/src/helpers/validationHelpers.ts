import * as Yup from 'yup'

const SignupSchema = Yup.object().shape({
    email: Yup.string()
        .email('Invalid email')
        .required('Required'),
    password: Yup.string()
        .required('Required'),
    confirmPassword: Yup.string()
        .oneOf([Yup.ref('password'), undefined], 'Passwords must match')
})

const LoginSchema = Yup.object().shape({
    email: Yup.string()
        .email('Invalid email')
        .required('Required'),
    password: Yup.string()
        .required('Required')
})

const SendEmailSchema = Yup.object().shape({
    email: Yup.string()
        .email('Invalid email')
        .required('Required')
})

const GeneralFormSchema = Yup.object().shape({
    name: Yup.string()
        .required('Required'),
    DoB: Yup.string()
        .required('Required'),
    gender: Yup.string()
        .required('Required'),
})

const ParentFormSchema = Yup.object().shape({
    invitationCode: Yup.string()
        .matches(/^[0-9]{6}$/, 'Invalid OTP')
        .required('Required'),
})

const StudentFormSchema = Yup.object().shape({
    email: Yup.string()
        .email('Invalid email')
        .required('Required'),
    password: Yup.string()
        .required('Required'),
    confirmPassword: Yup.string()
        .oneOf([Yup.ref('password'), undefined], 'Passwords must match'),

    name: Yup.string()
        .required('Required'),
    DoB: Yup.string()
        .required('Required'),
    gender: Yup.string()
        .required('Required'),
    teacher: Yup.string()
        .required('Required'),
    instrument: Yup.string()
        .required('Required'),
})

export {
    SignupSchema,
    LoginSchema,
    SendEmailSchema,
    GeneralFormSchema,
    ParentFormSchema,
    StudentFormSchema
}