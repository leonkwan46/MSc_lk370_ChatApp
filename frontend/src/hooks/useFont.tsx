import { useFonts } from 'expo-font'

const useFont = (font: string) => {
    const [fontLoaded, fontError] = useFonts({
        [font]: require('../assets/fonts/Lemon-Regular.ttf')
    })
    return { fontLoaded, fontError }
}

export default useFont