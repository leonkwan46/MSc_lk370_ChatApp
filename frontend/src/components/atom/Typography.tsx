import React, { type FC } from 'react'
import { Text, View } from 'react-native'
import { getColor, getFontSize } from '@/src/helpers/styleHelper'
import useFont from '@/src/hooks/useFont'


interface TypographyProps {
    children: string | React.ReactNode
    color?: string
    size?: string
    selected?: boolean
}

const Typography: FC<TypographyProps> = ({ children, color, size, selected }) => {
    const { fontLoaded } = useFont('Lemon-Regular')

    let selectedSize = getFontSize(size)
    let selectedColour = getColor(color)
    if (selected) selectedColour = getColor('primary')

    const textStyle = {
        fontFamily: 'Lemon-Regular',
        color: selectedColour,
        fontSize: selectedSize
    }
    
    return (
        <View>
            {fontLoaded ? (
                <Text style={textStyle}>{children}</Text>
            ) : (
                <Text>{children}</Text>
            )}
        </View>
    )
}

export default Typography

