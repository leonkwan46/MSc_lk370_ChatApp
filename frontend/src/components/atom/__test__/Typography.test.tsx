import React from 'react'
import renderer from 'react-test-renderer'
import Typography from '../Typography'
import useFont from '@/src/hooks/useFont'
import * as styleHelper from '@/src/helpers/styleHelper' 

// Mock the useFont hook
jest.mock('@/src/hooks/useFont', () => ({
    __esModule: true,
    default: jest.fn()
}))

describe('Typography Component', () => {
    let getColorSpy: jest.SpyInstance
    let getFontSizeSpy: jest.SpyInstance

    beforeEach(() => {
        // Spy on the getColor and getFontSize functions
        getColorSpy = jest.spyOn(styleHelper, 'getColor')
        getFontSizeSpy = jest.spyOn(styleHelper, 'getFontSize');
        
        // Reset mock implementations before each test
        (useFont as jest.Mock).mockReturnValue({ fontLoaded: true })
        getFontSizeSpy.mockReturnValue(15) // Default font size
        getColorSpy.mockReturnValue('black') // Default color
    })

    afterEach(() => {
        jest.resetAllMocks() // Reset all mocks after each test
    })

    it('renders correctly with default props', () => {
        const tree = renderer
            .create(<Typography>Default Text</Typography>)
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('applies custom styles when provided', () => {
        const customStyle = { color: 'green', fontSize: 25 }
        getFontSizeSpy.mockReturnValue(20) // Custom font size
        getColorSpy.mockReturnValue('blue') // Custom color

        const tree = renderer
            .create(
                <Typography extrasStyle={customStyle}>Styled Text</Typography>
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('uses primary color when selected prop is true', () => {
        getColorSpy.mockImplementation((color: string) => {
            if (color === 'primary') return '#D4AF37' // Example primary color
            return 'black' // Default color
        })

        const tree = renderer
            .create(<Typography selected={true}>Primary Color Text</Typography>)
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders correctly when font is not loaded', () => {
        (useFont as jest.Mock).mockReturnValue({ fontLoaded: false })

        const tree = renderer
            .create(<Typography>Fallback Text</Typography>)
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('uses default color if no color is provided', () => {
        const tree = renderer
            .create(<Typography>Default Color Text</Typography>)
            .toJSON()

        expect(tree).toMatchSnapshot()
    })

    it('renders correctly with size and color props', () => {
        getFontSizeSpy.mockReturnValue(25) // Custom font size
        getColorSpy.mockReturnValue('red') // Custom color

        const tree = renderer
            .create(
                <Typography size="xl" color="error">Error Color Text</Typography>
            )
            .toJSON()

        expect(tree).toMatchSnapshot()
    })
})
