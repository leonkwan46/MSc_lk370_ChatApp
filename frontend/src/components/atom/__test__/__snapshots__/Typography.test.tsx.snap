// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Typography Component applies custom styles when provided 1`] = `
<View>
  <Text
    style={
      {
        "color": "green",
        "fontSize": 25,
      }
    }
  >
    Styled Text
  </Text>
</View>
`;

exports[`Typography Component renders correctly when font is not loaded 1`] = `
<View>
  <Text
    style={null}
  >
    Fallback Text
  </Text>
</View>
`;

exports[`Typography Component renders correctly with default props 1`] = `
<View>
  <Text
    style={
      {
        "color": "black",
        "fontFamily": "Lemon-Regular",
        "fontSize": 15,
      }
    }
  >
    Default Text
  </Text>
</View>
`;

exports[`Typography Component renders correctly with size and color props 1`] = `
<View>
  <Text
    style={
      {
        "color": "red",
        "fontFamily": "Lemon-Regular",
        "fontSize": 25,
      }
    }
  >
    Error Color Text
  </Text>
</View>
`;

exports[`Typography Component uses default color if no color is provided 1`] = `
<View>
  <Text
    style={
      {
        "color": "black",
        "fontFamily": "Lemon-Regular",
        "fontSize": 15,
      }
    }
  >
    Default Color Text
  </Text>
</View>
`;

exports[`Typography Component uses primary color when selected prop is true 1`] = `
<View>
  <Text
    style={
      {
        "color": "#D4AF37",
        "fontFamily": "Lemon-Regular",
        "fontSize": 15,
      }
    }
  >
    Primary Color Text
  </Text>
</View>
`;
